### PR TITLE
Improve error message when a port is needed for odo link

### PR DESCRIPTION
**Pull request by geoand**
## What is the purpose of this change? What does it change?
See $SUBJECT

## Was the change discussed in an issue?
Fixes: #1174

## How to test changes?
Create a java backend component (which exposes multiple ports).
Create a frontend component.
Try to link the frontend to the backend. The error should now be what #1174 describes

### CODE REVIEW

**Comment by girishramnani** [comment]
@geoand https://github.com/redhat-developer/odo/blob/6055931a8ea917d93306351f7878e0ce06f604e0/pkg/secret/secret.go#L47 might also need to be changed

**Comment by geoand** [comment]
> @geoand
> 
> [odo/pkg/secret/secret.go](https://github.com/redhat-developer/odo/blob/6055931a8ea917d93306351f7878e0ce06f604e0/pkg/secret/secret.go#L47)
> 
> Line 47 in [6055931](/redhat-developer/odo/commit/6055931a8ea917d93306351f7878e0ce06f604e0)
>  Please select one of the following ports: '%s'`, componentName, port, strings.Join(availablePorts(secrets), ",")) 
> might also need to be changed

You're absolutely right, fixing now

**Comment by girishramnani** [comment]
a suggestion on the message going above 80 characters in one line, can we move a part to the next line ( not add a \n in the actual message ) just new line in the code.

**Comment by geoand** [comment]
> a suggestion on the message going above 80 characters in one line, can we move a part to the next line ( not add a \n in the actual message ) just new line in the code.

Makes sense, fixing now

File: pkg/secret/secret.go [hunk 1]
```diff
@@ -34,7 +34,8 @@ Please delete the component and recreate it using 'odo create'`, componentName)
 			return secrets[0].Name, nil
 		}
 		return "", fmt.Errorf(`Multiple secrets exist for component %s. 
-Please select one of the following ports: '%s'`, componentName, strings.Join(availablePorts(secrets), ","))
+Please select one of the following ports: '%s' 
+by supplying the --port option and rerun the command`, componentName, strings.Join(availablePorts(secrets), ","))
```
Comment (girishramnani) [inline pkg/secret/secret.go]: as you are using back ticks `` ` ``, the new line is actually not ignored and the `by supplying ...` prints in the next line.
Might wanna consider
 ```
"stuff " +
"next line" 
```

Comment (geoand) [reply pkg/secret/secret.go]: Hm... The message I am getting in the terminal  is the following:

```
None of the secrets that exist for component backend match port 9324. 
Please select one of the following ports: '8080' 
by supplying the --port option and rerun the command
```

 which looks correct. What am I missing here?

Comment (girishramnani) [reply pkg/secret/secret.go]: the wrapping of the message should be taken care by the terminal, try to resize the terminal once. This is what it looks like to me
image

Comment (girishramnani) [reply pkg/secret/secret.go]: so we shouldn't add a new line in the message but new line in the code so its more readable.

File: pkg/secret/secret.go [hunk 2]
```diff
@@ -33,8 +33,9 @@ Please delete the component and recreate it using 'odo create'`, componentName)
 		if len(secrets) == 1 {
 			return secrets[0].Name, nil
 		}
-		return "", fmt.Errorf(`Multiple secrets exist for component %s. 
-Please select one of the following ports: '%s'`, componentName, strings.Join(availablePorts(secrets), ","))
+		return "", fmt.Errorf("Multiple secrets exist for component %s. "+
+			"Please select one of the following ports: '%s' "+
```
Comment (mik-dass) [inline pkg/secret/secret.go]: Maybe something like
```
Please rerun the command by supplying the --port option with one of the following ports : `%s`
```
???

File: pkg/secret/secret.go [hunk 3]
```diff
@@ -43,9 +44,9 @@ Please select one of the following ports: '%s'`, componentName, strings.Join(ava
 			return secret.Name, nil
 		}
 	}
-	return "", fmt.Errorf(`None of the secrets that exist for component %s match port %s. 
-Please select one of the following ports: '%s'`, componentName, port, strings.Join(availablePorts(secrets), ","))
-
+	return "", fmt.Errorf("None of the secrets that exist for component %s match port %s. "+
+		"Please select one of the following ports: '%s' "+
```
Comment (mik-dass) [inline pkg/secret/secret.go]: Same as above

File: pkg/secret/secret.go (hunk 2 above)
Comment (geoand) [reply pkg/secret/secret.go]: Well, I just followed what was added in the open issue :)

Comment (mik-dass) [reply pkg/secret/secret.go]: That was just a suggestion :)

Comment (geoand) [reply pkg/secret/secret.go]: I don't mind either way of course :P

Comment (mik-dass) [reply pkg/secret/secret.go]: Ofcourse the decision is upto you :D

Comment (geoand) [reply pkg/secret/secret.go]: Let's go with this for the time being and if anyone complains we can always change it again

Comment (mik-dass) [reply pkg/secret/secret.go]: ok :+1:

**Comment by geoand** [comment]
@girishramnani One of the new integration tests seems to be flaky. Mind taking a look please?

**Comment by girishramnani** [comment]
Sure

**Comment by geoand** [comment]
Thanks!

**Pull request merged**